hepcrawl/utils.py

```diff
@@ -382,6 +384,49 @@ def get_licenses(
     return license
 
 
+def strict_kwargs(func):
+    """This decorator will disallow any keyword arguments that
+    do not begin with an underscore sign in the decorated method.
+    This is mainly to make errors while passing arguments to spiders
+    immediately visible. As we cannot remove kwargs from there altogether
+    (used by scrapyd), with this we can ensure that we are not passing unwanted
+    data by mistake.
+
+    Additionaly this will add all of the 'public' not-None kwargs to an
+    `_init_kwargs` field in the object for easier passing and all of the
+    arguments (including non-overloaded ones) to `_all_kwargs`.
+    (To make passing them forward easier.)
+
+    Args:
+        func (function): a spider method
+
+    Returns:
+        function: method which will disallow any keyword arguments that
+            do not begin with an underscore sign.
+    """
+    argspec = inspect.getargspec(func)
+    defined_arguments = argspec.args[1:]
+    spider_fields = ['settings']
+
+    allowed_arguments = defined_arguments + spider_fields
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        disallowed_kwargs = [
+            key for key in kwargs
+            if not key.startswith('_') and key not in allowed_arguments
+        ]
+
+        if disallowed_kwargs:
+            raise TypeError(
+                'Only underscored kwargs allowed in {}. '
+                'Check {} for typos.'.format(func, ', '.join(disallowed_kwargs))
```

Comment:
you should either:
* use `'%s' % var`
* make sure to handle unicode

As the format function will fail if one is unicode but the other is not (while the `%` will not). For example:
```
In [5]: '{}'.format(''.join([u'é']))
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-5-054f6ba4853c> in <module>()
----> 1 '{}'.format(''.join([u'é']))

UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 0: ordinal not in range(128)

In [6]: '%s' % ''.join([u'é'])
Out[6]: u'\xe9'
```

Reply:
Wouldn't it be a bit evil for someone to use unicode-named keyword arguments here?

Reply:
Evil indeed, but completely doable:
```
In [1]: kwargs = {'é': 123}

In [2]: def lolo(**kwargs):
   ...:     print kwargs
   ...:     

In [3]: lolo(**kwargs)
{'\xc3\xa9': 123}
```

This is more a 'good practice' than this specific case, I know you usually use format for sting interpolation ;)